x-pack/plugins/fleet/server/services/epm/kibana/assets/install.ts

```diff
@@ -121,6 +123,41 @@ export async function installKibanaAssets(options: {
 
   return installedAssets;
 }
+
+export async function installKibanaAssetsAndReferences({
+  savedObjectsClient,
+  savedObjectsImporter,
+  logger,
+  pkgName,
+  paths,
+  installedPkg,
+}: {
+  savedObjectsClient: SavedObjectsClientContract;
+  savedObjectsImporter: Pick<SavedObjectsImporter, 'import' | 'resolveImportErrors'>;
+  logger: Logger;
+  pkgName: string;
+  paths: string[];
+  installedPkg?: SavedObject<Installation>;
+}) {
+  const kibanaAssets = await getKibanaAssets(paths);
+  if (installedPkg) await deleteKibanaSavedObjectsAssets({ savedObjectsClient, installedPkg });
+  // save new kibana refs before installing the assets
+  const installedKibanaAssetsRefs = await saveKibanaAssetsRefs(
+    savedObjectsClient,
+    pkgName,
+    kibanaAssets
+  );
+
+  await installKibanaAssets({
```

Comment:
should there be any error handling here? e.g. if install assets fail, should the saved refs be cleaned up?

Reply:
I don’t believe it’s strictly necessary and none of the other asset types work this way. We only need these references so we can be sure to delete these assets during package upgrades or uninstalls. If they failed to be created, then nothing breaks and we can still uninstall the package.

I’m also trying to avoid changing too much behavior as part of this optimization.

Reply:
Sorry to clarify: all asset types create their references first and don’t clean them up if the asset fails to actually be installed, so this is consistent with the others.